File: backTracking/subSets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        cur = []
        numLen = len(nums)
        
        def dfs(ind):
            #base case
            if ind >= numLen:
                res.append(list(cur))
                return 
            
            #decision to add and forward track
            cur.append(nums[ind])
            dfs(ind + 1)

            #backtrack then make decision not to add
            cur.pop()
            dfs(ind + 1)
        
        #init/call depth first solution search function and then return
        dfs(0)    
        return res

File: backTracking/subSets.cpp
class Solution {
public:
    void bt(vector<int>& nums, vector<int>& cur_res, int index, vector<vector<int>>& res){
        //base case: we are at the end of our nums array or index == nums.size()
        if(index == nums.size()){
            res.push_back(cur_res);
            return;
        }

        cur_res.push_back(nums[index]);
        bt(nums, cur_res, index + 1, res);
        cur_res.pop_back();
        bt(nums, cur_res, index + 1, res);  
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res = {};
        vector<int> cur_res = {};
        int index = 0;
        bt(nums, cur_res, index, res);
        return res;
    }
};
